### PR TITLE
[clang-format] revert to string << string handling to previous default

### PR DESCRIPTION
**Pull request by mydeveloperday**
Fixes 88433

A change made to the handling of chevron operators causes a large amount of flux in code bases that were previously using clang-format, this fix reverts that change to the default behaviour but adds that new behaviour behind a new option.

### CODE REVIEW

**Comment by mydeveloperday** [comment]
Fixes #88483

**Comment by mydeveloperday** [comment]
Other related issues

#44363 , #59797

File: clang/lib/Format/TokenAnnotator.cpp [hunk 1]
```diff
@@ -5598,10 +5598,34 @@ bool TokenAnnotator::mustBreakBefore(const AnnotatedLine &Line,
...
+  }
+  if (Style.BreakChevronOperator == FormatStyle::BCOS_BetweenNewlineStrings) {
+    if (Right.is(tok::lessless) && Right.Next &&
+        Right.Next->is(tok::string_literal) && Left.is(tok::string_literal) &&
+        Left.TokenText.ends_with("\\n\"")) {
+      return true;
+    }
+  }
+  if (Style.BreakChevronOperator == FormatStyle::BCOS_Always) {
+    // can be std::os or os
```
Comment (HazardyKnusperkeks) [inline clang/lib/Format/TokenAnnotator.cpp:5615]: I don't understand that comment.

File: clang/lib/Format/TokenAnnotator.cpp [hunk 2]
```diff
@@ -5598,10 +5598,34 @@ bool TokenAnnotator::mustBreakBefore(const AnnotatedLine &Line,
...
+    auto *FirstChevron = Right.Previous;
+    while (FirstChevron) {
```
Comment (HazardyKnusperkeks) [inline clang/lib/Format/TokenAnnotator.cpp:5617]: Is there a test case which utilizes this loop?

File: clang/include/clang/Format/Format.h [hunk 3]
```diff
@@ -2193,6 +2193,41 @@ struct FormatStyle {
...
+    ///      << "bbbbb"
+    ///      << "ccccc\n"
+    ///      << "\n";
+    /// \endcode
+    BCOS_Always
+  };
+
+  /// Break Between Chevron Operators
+  /// \version 19
+  BreakChevronOperatorStyle BreakChevronOperator;
```
[the remark this answers is not in the document]
Comment (mydeveloperday) [reply clang/include/clang/Format/Format.h:2229]: So later commit in the PR do >> as well as << so maybe just BreakStreamOperator? WDYT

File: clang/lib/Format/TokenAnnotator.cpp (hunk 2 above)
Comment (mydeveloperday) [reply clang/lib/Format/TokenAnnotator.cpp:5617]: yes anything with std::os in

File: clang/lib/Format/TokenAnnotator.cpp (hunk 1 above)
Comment (mydeveloperday) [reply clang/lib/Format/TokenAnnotator.cpp]: when you have 

os <<  "A" 
     << "B"

you don't want to break between os and << so the loop tracks back to determine if this is the first << in the line, this must also handle std::os << and a::b::os etc..  whats its really saying is don't break before the first <<.

File: clang/lib/Format/TokenAnnotator.cpp [hunk 4]
```diff
@@ -5598,11 +5598,45 @@ bool TokenAnnotator::mustBreakBefore(const AnnotatedLine &Line,
...
+    if (Right.is(tok::lessless) && Right.Next &&
+        Right.Next->is(tok::string_literal) && Left.is(tok::string_literal) &&
+        Left.TokenText.ends_with("\\n\"")) {
+      return true;
+    }
+    break;
+  }
+  case FormatStyle::BCOS_Always: {
+    // Don't break after the very first << or >>
+    // but the Left token can be os or std::os so
```
Comment (HazardyKnusperkeks) [inline clang/lib/Format/TokenAnnotator.cpp:5619]: Still don't understand the part of `os` or `std::os`.

File: clang/include/clang/Format/Format.h (hunk 3 above)
[the remark this answers is not in the document]
Comment (HazardyKnusperkeks) [reply clang/include/clang/Format/Format.h]: That's good.

File: clang/lib/Format/TokenAnnotator.cpp (hunk 1 above)
Comment (HazardyKnusperkeks) [reply clang/lib/Format/TokenAnnotator.cpp]: Yeah ok, but I'd skip the `os`.

**Comment by owenca** [comment]
Please see https://github.com/llvm/llvm-project/issues/88483#issuecomment-2053928993 and Conversation in #69859, which was the first attempt at adding an option for breaking consecutive stream insertion operations. Instead of handling everything specified in https://github.com/llvm/llvm-project/pull/69859#issuecomment-1776064714 in a single patch, we can take a few steps:

1. Handle when to break _before_ `<<`. The current default (since 18.x) is only when the left operand is a string literal ending with a newline and the right operand is a string literal.
2. Handle whether to indent or align after the break.
3. Do the above for breaking breaking _after_ `<<`.

(Ditto for stream extraction operations if needed.)

For No. 1 above, an `enum` option (e.g. `BreakBeforeStreamInsertionOperator`) with `Leave`, `AfterNewline` (LLVM default), `Always`, and `Never` may be sufficient. @s1Sharp

**Pull request closed without merging**